rust/agama-server/src/web/common.rs

```diff
@@ -108,3 +116,120 @@ async fn build_service_status_proxy<'a>(
         .await?;
     Ok(proxy)
 }
+
+/// Builds a router to the `org.opensuse.Agama1.Progress`
+/// interface of the given D-Bus object.
+///
+/// ```no_run
+/// # use axum::{extract::State, routing::get, Json, Router};
+/// # use agama_lib::connection;
+/// # use agama_server::web::common::progress_router;
+/// # use tokio_test;
+///
+/// # tokio_test::block_on(async {
+/// async fn hello(state: State<HelloWorldState>) {};
+///
+/// #[derive(Clone)]
+/// struct HelloWorldState {};
+///
+/// let dbus = connection().await.unwrap();
+/// let progress_router = progress_router(
+///   &dbus, "org.opensuse.HelloWorld", "/org/opensuse/hello"
+/// ).await.unwrap();
+/// let router: Router<HelloWorldState> = Router::new()
+///   .route("/hello", get(hello))
+///   .merge(progress)
+///   .with_state(HelloWorldState {});
+/// });
+/// ```
+///
+/// * `dbus`: D-Bus connection.
+/// * `destination`: D-Bus service name.
+/// * `path`: D-Bus object path.
+pub async fn progress_router<T>(
+    dbus: &zbus::Connection,
+    destination: &str,
+    path: &str,
+) -> Result<Router<T>, ServiceError> {
+    let proxy = build_progress_proxy(dbus, destination, path).await?;
+    let state = ProgressState { proxy };
+    Ok(Router::new()
+        .route("/progress", get(progress))
+        .with_state(state))
+}
+
+#[derive(Clone)]
+struct ProgressState<'a> {
+    proxy: ProgressProxy<'a>,
+}
+
+async fn progress(State(state): State<ProgressState<'_>>) -> Json<Progress> {
+    let proxy = state.proxy;
+    let progress = Progress::from_proxy(&proxy).await.unwrap();
+    Json(progress)
+}
+
+#[pin_project]
+pub struct ProgressStream<'a> {
+    #[pin]
+    inner: PropertyStream<'a, (u32, String)>,
+    proxy: ProgressProxy<'a>,
+}
+
+pub async fn progress_stream<'a>(
+    dbus: zbus::Connection,
+    destination: &'static str,
+    path: &'static str,
+) -> ProgressStream<'a> {
+    let proxy = build_progress_proxy(&dbus, destination, path)
+        .await
+        .unwrap();
+    ProgressStream::new(proxy).await
+}
+
+impl<'a> ProgressStream<'a> {
+    pub async fn new(proxy: ProgressProxy<'a>) -> Self {
+        let stream = proxy.receive_current_step_changed().await;
+        ProgressStream {
+            inner: stream,
+            proxy,
+        }
+    }
+}
+
+impl<'a> Stream for ProgressStream<'a> {
```

Comment:
own stream? nice

Reply:
Yes, it allows us to keep some additional state.